dap_prinz_green_jobs/pipeline/green_measures/occupations/soc_map.py

```diff
@@ -0,0 +1,378 @@
+        There are additional words to include in the job title if at first
+        it is not unique.
+        """
+
+        col_name_0 = "INDEXOCC NATURAL WORD ORDER"
+        col_name_1 = "ADD"
+        col_name_2 = "IND"
+
+        jobtitle_soc_data[f"{col_name_0} and {col_name_1}"] = jobtitle_soc_data.apply(
+            lambda x: x[col_name_0] + " " + x[col_name_1]
+            if pd.notnull(x[col_name_1])
+            else x[col_name_0],
+            axis=1,
+        )
+        jobtitle_soc_data[
+            f"{col_name_0} and {col_name_1} and {col_name_2}"
+        ] = jobtitle_soc_data.apply(
+            lambda x: x[f"{col_name_0} and {col_name_1}"] + " " + x[col_name_2]
+            if pd.notnull(x[col_name_2])
+            else x[f"{col_name_0} and {col_name_1}"],
+            axis=1,
+        )
+
+        # Try to find a unique job title to SOC 2020 4 or 6 code mapping
+        job_title_2_soc6_4 = {}
+        for job_title, grouped_soc_data in jobtitle_soc_data.groupby(col_name_0):
+            if grouped_soc_data["SOC_2020_EXT"].nunique() == 1:
+                job_title_2_soc6_4[job_title] = (
+                    grouped_soc_data["SOC_2020_EXT"].unique()[0],
+                    grouped_soc_data["SOC_2020"].unique()[0],
+                )
+            else:
+                for job_title_1, grouped_soc_data_1 in grouped_soc_data.groupby(
+                    f"{col_name_0} and {col_name_1}"
+                ):
+                    if grouped_soc_data_1["SOC_2020_EXT"].nunique() == 1:
+                        job_title_2_soc6_4[job_title_1] = (
+                            grouped_soc_data_1["SOC_2020_EXT"].unique()[0],
+                            grouped_soc_data_1["SOC_2020"].unique()[0],
+                        )
+                    else:
+                        for (
+                            job_title_2,
+                            grouped_soc_data_2,
+                        ) in grouped_soc_data_1.groupby(
+                            f"{col_name_0} and {col_name_1} and {col_name_2}"
+                        ):
+                            if grouped_soc_data_2["SOC_2020_EXT"].nunique() == 1:
+                                job_title_2_soc6_4[job_title_2] = (
+                                    grouped_soc_data_2["SOC_2020_EXT"].unique()[0],
+                                    grouped_soc_data_2["SOC_2020"].unique()[0],
+                                )
+
+        return job_title_2_soc6_4
+
+    def embed_texts(
+        self,
+        texts: list,
+    ) -> np.array(object):
+        logger.info(f"Embedding texts in {len(texts)/self.batch_size} batches")
+        all_embeddings = []
+        for batch_texts in tqdm(chunks(texts, self.batch_size)):
+            all_embeddings.append(
+                self.bert_model.encode(np.array(batch_texts), batch_size=32)
+            )
+        all_embeddings = np.concatenate(all_embeddings)
+
+        return all_embeddings
```

Comment:
there's the `bert_vectoriser.py` BertModel class in utils you could use

Reply:
it's true - but I wanted to keep this script slightly independent from the rest of the repo so we could move it out at some point? do you think this I should change it? maybe it is a bit messy to have both!

Reply:
liz - (maybe) to do

Reply:
actually really good point - ignore me!